webapp/src/RepoGit.ts

```diff
@@ -0,0 +1,147 @@
+import { simpleGit, SimpleGit, SimpleGitOptions } from 'simple-git';
+import { WriteLanguageFileError, WriteLanguageFileErrors } from '@/errors';
+
+export class RepoGit {
+  private readonly git: SimpleGit;
+  private lyraConfig?: LyraConfig;
+
+  constructor(public readonly repoPath: string) {
+    const options: Partial<SimpleGitOptions> = {
+      baseDir: repoPath,
+      binary: 'git',
+      maxConcurrentProcesses: 1,
+      trimmed: false,
+    };
+    this.git = simpleGit(options);
+  }
+
+  /**
+   * Checkout base branch and pull
+   * @returns base branch name
+   */
+  public async checkoutBaseAndPull(): Promise<string> {
+    const lyraConfig = await this.getLyraConfig();
+    await this.git.checkout(lyraConfig.baseBranch);
+    await this.git.pull();
+    return lyraConfig.baseBranch;
+  }
+
+  public async saveLanguageFiles(projectPath: string): Promise<string[]> {
+    const lyraConfig = await this.getLyraConfig();
+    const projectConfig = lyraConfig.getProjectConfigByPath(projectPath);
+    const projectStore = await Cache.getProjectStore(projectConfig);
+    const languages = await projectStore.getLanguageData();
+
+    return await this.writeLangFiles(
+      languages,
+      projectConfig.absTranslationsPath,
+    );
+  }
+
+  public async statusChanged(): Promise<boolean> {
+    const status = await this.git.status();
+    return status.files.length > 0;
+  }
+
+  public async newBranchCommitAndPush(
+    branchName: string,
+    addFiles: string[],
+    commitMsg: string,
+  ): Promise<void> {
+    const lyraConfig = await this.getLyraConfig();
+    await this.git.checkoutBranch(branchName, lyraConfig.baseBranch);
+    await this.git.add(addFiles);
+    await this.git.commit(commitMsg);
+    await this.git.push(['-u', 'origin', branchName]);
+  }
+
+  public async createPR(
+    branchName: string,
+    prTitle: string,
+    prBody: string,
+    githubOwner: string,
+    githubRepo: string,
+    githubToken: string,
+  ): Promise<string> {
+    const lyraConfig = await this.getLyraConfig();
+    const octokit = new Octokit({
+      auth: githubToken,
+      baseUrl: 'https://api.github.com',
+      log: {
+        debug: debug,
+        error: () => {},
+        info: info,
+        warn: warn,
+      },
+      request: {
+        agent: undefined,
+        fetch: undefined,
+        timeout: 0,
+      },
+      timeZone: Intl.DateTimeFormat().resolvedOptions().timeZone,
+      userAgent: 'Lyra v' + packageJson.version,
+    });
+
+    const response = await octokit.rest.pulls.create({
+      base: lyraConfig.baseBranch,
+      body: prBody,
+      head: branchName,
+      owner: githubOwner,
+      repo: githubRepo,
+      title: prTitle,
+    });
+
+    return response.data.html_url;
+  }
+
+  private async getLyraConfig(): Promise<LyraConfig> {
+    if (this.lyraConfig) {
+      return this.lyraConfig;
+    }
+    this.lyraConfig = await LyraConfig.readFromDir(this.repoPath);
```

Comment:
Note that while one invocation is awaiting readFromDir, another invocation can find lyraConfig false-like and attempt its own readFromDir.

We do not have any concurrent usage of RepoGit instances so there is no bug now. Even if we did introduce concurrent use later, rereading it will be a non-problem as long as the configuration file has not changed on disk.

I think the risk is low enough that we can merge this as is and fix it in a later issue instead.

Reply:
Agree the that risk is very low, and we can create a separate issue for it "make readFromDir Atomic" that guarantee that reading lyra config file of the same repo (same lyra.yaml file) does not occur simultaneously